[hendo.park] 지하철 서비스 step3 코드리뷰 부탁드립니다.

hendo-park opened the pull request:
안녕하세요, hendo입니다..!

우선 항상 알찬 리뷰에 감사드립니다..! 

질문이 2가지 있습니다.

1.
    private int makeFare(ShortestGraph shortestGraph, LoginMember loginMember) {
        int fare = INITIAL_FARE;
        fare = applyFareStrategy(new DistancePolicy(fare, shortestGraph.getGraphPath()));
        fare = applyFareStrategy(new LinePolicy(fare, shortestGraph.getGraphPath()));
        fare = applyFareStrategy(new MemberPolicy(fare, loginMember));
        return fare;
    }

Fare Class 의 코드중 일부입니다. (fare을 일급 객체로 만들었습니다)

Fare을 계산하기 위한 로직이 담긴 코드인데요..! 요금정책과 같은 정책은 스트레테지패턴을 적용보는게 어떨까 싶어서 적용 해 보았으나..
실상 연속적으로 적용되어야 하는 부분이기 때문에 크게 유의미한 패턴이 되지 못한 것 같습니다..!
리뷰어님께서는 이부분에 대해 어떻게 하면 더 명확하게 명시되면서도 좋은 패턴혹은 방벙을 사용하셨을지 궁금합니다..!

2.
NullCheck 라고 해야할까요? Exception 처리에 대한 부분이 어디서, 언제 이루어져야할지가 계속해서 햇갈립니다..!
이부분도 꼼꼼한 테스트 작성에 들어가는걸까요?
어디서 꼭 해줘야할지라던가, 이부분은 널이 들어올 가능성이 거의 없어보이는데 꼭 해줘야하나 이런 부분도 많은 것 같습니다..!
혹시 기준이나 팁이 있다면 알려주신다면 감사하겠습니다..!

Commit:
feat : step1 임시구현

Commit:
feat : step1 임시구현

Commit:
Changes by kuri

Commit:
feat : step1 구현

Commit:
feat : step2 구현

Commit:
feat : line 정보를 가진 Weight 생성

Commit:
feat : step3 요금정책 구현

Commit:
refactor : Path 부분 리팩토링

Commit:
refactor : Path 부분 리팩토링 (2)

객체의 책임을 다시한번 나누고, 리팩토링을 다시 진행하였습니다.

Commit:
refactor : 전체적인 리팩토링 최종 진행 및 테스트 작성

리팩토링을 확인하고 테스트를 작성하였습니다.

dongjoon1251 commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,40 @@
+
+    public TokenResponse createAuth(String email, String password) {
+        if(!memberDao.findByEmail(email).getPassword().equals(password)){
+            throw new InvalidMemberException("이메일이 없거나, 패스워드가 일치하지 않음");
```
커스텀 예외 클래스 👍 
에러메시지가 항상 같다면 예외클래스 내부에 상수로 가지게하면 어떨까요?

dongjoon1251 commented on src/main/java/subway/auth/ui/AuthController.java:
```diff
@@ -1,5 +1,30 @@
+
+@RestController
 public class AuthController {
     // TODO: 로그인(토큰 발급) 요청 처리하기
```
불필요한 주석은 제거해주세요 :)

dongjoon1251 commented on src/main/java/subway/auth/ui/AuthController.java:
```diff
@@ -1,5 +1,30 @@
+    }
+
+    @PostMapping(value = "/login/token", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<TokenResponse> createAuth(@RequestBody Map<String, String> params){
```
Map을 요청 body로 받기도 하지만 DTO로 따로 만들어서 받아보면 어떨까요?

dongjoon1251 commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -32,13 +32,14 @@ public Line insert(Line line) {
 
     public Line findById(Long id) {
-        String sql = "select L.id as line_id, L.name as line_name, L.color as line_color, " +
+        String sql = "select L.id as line_id, L.name as line_name, L.color as line_color, L.extra_fare as line_extra_fare, " +
```
쿼리도 상수로 분리해보면 어떨까요? :)

dongjoon1251 commented on src/main/java/subway/path/domain/Fare.java:
```diff
@@ -0,0 +1,35 @@
+        int fare = INITIAL_FARE;
+        fare = applyFareStrategy(new DistancePolicy(fare, shortestGraph.getGraphPath()));
+        fare = applyFareStrategy(new LinePolicy(fare, shortestGraph.getGraphPath()));
+        fare = applyFareStrategy(new MemberPolicy(fare, loginMember));
```
전략 패턴을 적용해주셨네요 👍 
여러개의 전략이 순차적으로 적용되는 것으로 보입니다. 구현해주신 부분과 함께 데코레이터 패턴을 적용해보면 어떨까합니다. :)

dongjoon1251 reviewed (changes requested):
안녕하세요. 리뷰를 맡게된 제노스입니다.
코드를 잘 작성해주시고, 클래스도 잘 분리해주셨네요 👍 
Policy를 적용한 부분도 인상깊었습니다.
추가적인 코멘트를 드리자면 다음과 같습니다. 
1. 도메인과 관련된 테스트 코드를 추가해주세요 
2. service 레이어에 Request DTO를 그대로 전달하기보다 파라미터가 많아지더라도 필요한 필드만 전달해주세요
3. 불필요한 주석은 제거해주세요 ex)  TODO: 

마지막으로 질문에 대한 답변은 다음과 같습니다.
1. 정책이라는 도메인을 잘 분리해주셨습니다. 공통적으로 필요한 기능을 추출해보려는 시도가 아조 보기 좋았습니다. 더 명확한 패턴 적용에 대한 정답은 없습니다. 적용해주신 방법도 좋은 방법으로 보입니다. 연속으로 같은 기능이 적용된다는 점에서 구현해주신 방법과 더불어서 데코레이터 패턴을 적용해볼 수 있을 것 같습니다. 
2. 사용자가 전달하는 요청 DTO에서는 null 체크를 피할 수 없습니다. 하지만 null 체크가 DTO를 처음 받는 컨트롤러나 DTO내부에서 처리된다면 어떨까요? null을 체크해아 하는 부분을 한정지을 수 있습니다. 그래서 null이 발생할 수 있는 초기 부분만 잘 체크하고 나머지 부분에서는 불필요하게 null 체크를 하지 않아도 좋습니다. 꼭 필요하다면 @NotNull 과 같은 애노테이션을 메소드 파라미터에 추가하는 경우도 많습니다 :)

Commit:
refactor : Service가 DTO객체를 받지 않도록함. 주석제거.

Commit:
refactor : 에러 메세지 상수화

Commit:
refactor : sql문 상수화

Commit:
refactor : 컨트롤러에서 요청은 DTO로 받도록 수정

Commit:
refactor : 전략패턴을 데코레이터패턴으로 변경

Commit:
feat : 도메인에 따른 테스트 추가

Commit:
refactor : favorite 리팩토링

Commit:
refactor : 널 체크 추가

input쪽 dto 부분에 널체크를 시행하고, Optional 키워드를 통해 dao 에서 에러 처리를 하는 것이 아닌 최소 서비스 단에서 반환하도록 변환.

Chaeoon-Park commented:
안녕하세요 제노스! 
어제 회사에 다녀와서 이거 저거 하느라고 바꾼것에 비해서 리뷰 요청이 늦었네요 죄송해요..! ㅠㅠ
말씀해주신 부분에 대해 모두 적용해보자 노력하고, favorite부분도 손을 조금 보았는데,
특히 널체크 부분이 제대로 한게 맞는지 잘 모르겠네요 ㅠㅠ
정성스러운 리뷰 정말 감사드립니다!

dongjoon1251 commented on src/main/java/subway/auth/dto/TokenRequest.java:
```diff
@@ -8,15 +10,26 @@ public TokenRequest() {
     }
 
+    private void validInput(String email, String password) {
+        if (email == null || email.trim().isEmpty()) {
```
ObjectUtils.isEmpty 를 적용해보셔도 좋을 것 같습니다. trim() 부분이 있어서 좀 애매하긴 하네요.

dongjoon1251 commented on src/main/java/subway/line/dto/LineRequest.java:
```diff
@@ -28,6 +36,25 @@ public LineRequest(String name, String color, Long upStationId, Long downStation
     }
 
+    private void validInput(String name, String color, Long upStationId, Long downStationId, int distance) {
+        if (name == null || name.trim().isEmpty()) {
```
null이거나 trim() 된 String이 empty인지 확인하는 코드가 중복되네요. Util의 static method로 분리시켜봐도 좋을 것 같습니다.

dongjoon1251 reviewed (approved):
피드백을 잘 반영해주셨네요 :)
코멘트 몇가지를 코드상에 남겼습니다. 다음 미션을 진행하실 때 고려해주세요. 
이번 미션은 여기서 마무리하도록 하겠습니다.
고생하셨습니다!
